---
title: "**phenomis**: Phenomics data analysis"
author: "Etienne A. Thevenot"
date: "`r doc_date()`"
package: "`r pkg_ver('phenomis')`"

vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: "phenomis.bib"
output:
  BiocStyle::html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: false
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width = 8,
                      fig.height = 8,
                      fig.path = 'figures/temp/')
```

# Introduction

## Context

Analysis of a phenomics (e.g. metabolomics) data set (i.e. samples times variables table of peak or bucket intensities generated by preprocessing tools such as XCMS) is aimed at **mining the data** (e.g. trends and outliers) and detecting features of predictive value (**biomarker discovery**). It comprises multiple steps including:

* **Post-processing** (quality control, normalization and/or transformation of intensities)

* **Statistical analysis** (univariate hypothesis testing, multivariate modeling, feature selection)

* **Annotation** (physico-chemical properties, biological pathways)

The **phenomis** package addresses the two first steps, and can be combined with other packages for multivariate modeling, such as the [**ropls**](https://doi.org/10.18129/B9.bioc.ropls) and [**biosigner**](https://doi.org/10.18129/B9.bioc.biosigner) Bioconductor packages, as described below.

## Methods

![Methods from the **phenomis**, **ropls**, and **biosigner** packages for the analysis of metabolomics datasets; specific parameter values used for the **sacurine** dataset described in the 'Hands-on' part below are provided as examples.](figures/permanent/phenomis_methods.png)

## Formats

### 3 tabular file format used for import/export

Input (i.e. preprocessed) data consists of a 'samples times variables' matrix of intensities (**datMatrix** numeric matrix), in addition to sample and variable metadata (**sampleMetadata** and **variableMetadata** data frames). Theses 3 tables can be conveniently imported to/exported from R as tabular files:

![3 table format used as input/output from the data analysis workflow.](figures/permanent/phenomis_3table_format.png)

### **ExpressionSet** class used within the data analysis workflow

Within the R workflow, the **ExpressionSet** class perfectly handles these 3 tables (for additional information about *ExpressionSet* class, see the ['An introduction to Biobase and ExpressionSets'](https://bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf) documentation from the [**Biobase**](https://doi.org/doi:10.18129/B9.bioc.Biobase) package). The following table describes useful Biobase methods for the handling of ExpressionSet objects: 

| Biobase methods | Description |
|---|---|
| **exprs(eset)** | 'variable times samples' numeric matrix - dataMatrix |
| **pData(eset)** | sample metadata data frame - sampleMetadata |
| **fData(eset)** | variable metadata data frame - variableMetadata |
| **sampleNames(eset)** | sample names |
| **featureNames(eset)** | variable names |
| **dims(eset)** | 2-element numeric vector of 'Features' and 'Samples' dimensions |
| **varLabels(eset)** | Column names of the sampleMetadata, pData(eset) |
| **fvarLabels(eset)** | Column names of the variableMetadata, fData(eset) |

### **MultiDataSet** class for multiple datasets

All *phenomis* methods (including *reading* and *writing*) can also be applied to multiple omics datasets: in this case, the **MultiDataSet** class is used instead of *ExpressionSet* [[@hernandez-ferrer_multidataset:_2016]](https://doi.org/10.1186/s12859-016-1455-1).

## Text and graphical outputs

Text and graphics can be handled with the *phenomis* methods by setting the two arguments:

* **report.c**: if set to 'interactive' [default], messages are displayed interactively; if set to a character corresponding the a filename (with the '.txt' extension), messages are diverted to this file by using the *sink* function internally (the same file can be appended by successive methods); if set to 'none', messages are suppressed

* **figure.c**: if set to 'interactive' [default], graphics are displayed interactively; if set to a character corresponding the a filename (with the '.pdf' extension), a pdf file with the plot is generated instead; if set to 'none', graphics are suppressed

## Availability

The **phenomis** package can be installed from github with:
```{r installation, eval = FALSE}
devtools::install_github("ethevenot/phenomis")
```

# Hands-on

## The **sacurine** cohort study

As an example, we will use the **phenomis** package to study the **sacurine** human cohort. The study is aimed at characterizing the physiological variations of the human urine metabolome with age, body mass index (BMI), and gender [[@thevenot_analysis_2015]](https://doi.org/10.1021/acs.jproteome.5b00354). Urine samples from 184 volunteers were analyzed by reversed-phase (C18) ultrahigh performance liquid chromatography (UPLC) coupled to high-resolution mass spectrometry (LTQ-Orbitrap). Raw data are publicly available on the MetaboLights repository ([MTBLS404](https://www.ebi.ac.uk/metabolights/MTBLS404)).

This vignette describes the statistical analysis of the data set from the negative ionization mode (113 identified metabolites at MSI levels 1 or 2):

* **correcting**: Correction of the signal drift by local regression
(loess) modeling of the intensity trend in pool samples [[@dunn_procedures_2011]](https://doi.org/10.1038/nprot.2011.335); Adjustment of oﬀset diﬀerences between the two analytical
batches by using the average of the pool intensities in each batch
[[@van_der_kloet_analytical_2009]](https://doi.org/10.1021/pr900499r)

* Variable quality control by discarding features
with a coeﬃcient of variation above 30% in *pool* samples

* Normalization by the sample osmolality

* **transforming**: log10 transformation

* **inspecting**: Computing metrics to filter out outlier samples according to the Weighted Hotellings’T2 distance [[@tenenhaus_approche_1999]](http://www.numdam.org/item/?id=RSA_1999__47_2_5_0), the Z-score of one
of the intensity distribution deciles [[@alonso_astream:_2011]](https://doi.org/10.1093/bioinformatics/btr138), and the Z-score
of the number of missing values [[@alonso_astream:_2011]](https://doi.org/10.1093/bioinformatics/btr138). A 0.001 threshold for all p-values results in the HU_096 sample being discarded

* **hypotesting**: Univariate hypothesis testing of significant variations with age, BMI, or between genders (Student's T test with Benjamini Hochberg correction)

* PCA exploration of the data set; [**ropls**](https://doi.org/10.18129/B9.bioc.ropls) Bioconductor package [[@thevenot_analysis_2015]](https://doi.org/10.1021/acs.jproteome.5b00354)

* **clustering**: Hierarchical clustering

* OPLS(-DA) modeling of age, BMI and gender; [**ropls**](https://doi.org/10.18129/B9.bioc.ropls) Bioconductor package [[@thevenot_analysis_2015]](https://doi.org/10.1021/acs.jproteome.5b00354)

* Selection of the features which significantly contributes to the discrimination between gender with PLS-DA, Random Forest, or Support Vector Machines classifiers; [**biosigner**](https://doi.org/10.18129/B9.bioc.biosigner) Bioconductor package [[@rinaudo_biosigner:_2016]](https://doi.org/10.3389/fmolb.2016.00026)

A Galaxy version of this analysis is available [W4M00001 'Sacurine-statistics'](https://doi.org/10.15454/1.4811121736910142E12) on the [Workflow4metabolomics.org](https://workflow4metabolomics.org) online infrastructure [[@guitton_create_2017]](https://doi.org/10.1016/j.biocel.2017.07.002)


## **reading**: Reading the data

The **reading** function reads the data sets and builds the *ExpressionSet* object. For additional information about *ExpressionSet* class, see the "An introduction to Biobase and ExpressionSets" documentation from the [Biobase](https://doi.org/doi:10.18129/B9.bioc.Biobase) package.


```{r reading}
sacurine.eset <- phenomis::reading(system.file("extdata/sacurine", package = "phenomis"))
```

## **inspecting**: Looking at the data

```{r inspecting}
sacurine.eset <- phenomis::inspecting(sacurine.eset)
```

## Post-processing

### **correcting**: Correcting signal drift and batch effect

```{r correcting}
sacurine.eset <- phenomis::correcting(sacurine.eset,
                                      reference.c = "pool",
                                      batch.c = "batch",
                                      injection_order.c = "injectionOrder",
                                      sample_type.c = "sampleType")
```

### Variable filtering

* using **inspecting** to compute the 'pool_CV' metric
```{r inspecting_variable_filtering}
sacurine.eset <- phenomis::inspecting(sacurine.eset)
```

* discarding features with 'pool_CV' > 0.3
```{r discarding_pool_CV_sup_0.3}
sacurine.eset <- sacurine.eset[Biobase::fData(sacurine.eset)[, "pool_CV"] <= 0.3, ]
```

* discarding the 'pool' observations
```{r discarding_pools}
sacurine.eset <- sacurine.eset[, Biobase::pData(sacurine.eset)[, "sampleType"] != "pool"]
print(sacurine.eset)
```


### Normalizing

```{r normalizing_osmolality}
Biobase::exprs(sacurine.eset) <- sweep(Biobase::exprs(sacurine.eset),
                                       2,
                                       Biobase::pData(sacurine.eset)[, "osmolality"],
                                       "/")
```

### **transforming**: Transforming the data intensities

```{r transforming}
sacurine.eset <- phenomis::transforming(sacurine.eset, method.c = "log10")

```

### Sample filtering

```{r sample_filtering}
sacurine.eset <- phenomis::inspecting(sacurine.eset)
sacurine.eset <- sacurine.eset[, Biobase::pData(sacurine.eset)[, "hotel_pval"] >= 0.001 &
                                 Biobase::pData(sacurine.eset)[, "miss_pval"] >= 0.001 &
                                 Biobase::pData(sacurine.eset)[, "deci_pval"] >= 0.001]
```

Final visual check of the data before performing the statistics

```{r final_check}
phenomis::inspecting(sacurine.eset)
```

## **hypotesting**: Univariate hypothesis testing

```{r hypotesting}
sacurine.eset <- phenomis::hypotesting(sacurine.eset,
                                       test.c = "ttest",
                                       factor_names.vc = "gender",
                                       adjust.c = "BH",
                                       adjust_thresh.n = 0.05)

```

## Unsupervised analysis

### Principal component analysis

[**ropls**](https://doi.org/10.18129/B9.bioc.ropls) Bioconductor package (already loaded as a dependance from *phenomis*) [[@thevenot_analysis_2015]](https://doi.org/10.1021/acs.jproteome.5b00354)

```{r PCA}
sacPca <- ropls::opls(sacurine.eset, info.txt = NA)
ropls::plot(sacPca,
            parAsColFcVn = Biobase::pData(sacurine.eset)[, "gender"],
            typeVc = "x-score")
ropls::plot(sacPca,
            parAsColFcVn = Biobase::pData(sacurine.eset)[, "age"],
            typeVc = "x-score")
ropls::plot(sacPca,
            parAsColFcVn = Biobase::pData(sacurine.eset)[, "bmi"],
            typeVc = "x-score")
```

```{r pca_getEset}
sacurine.eset <- ropls::getEset(sacPca)
```

### **clustering**: hierarchical clustering

```{r heatmap}
sacurine.eset <- phenomis::clustering(sacurine.eset, correl.c = "spearman",
                                      clusters.vi = c(5, 3))
```

## Supervised modeling

### (O)PLS(-DA) modeling

With the [**ropls**](https://doi.org/10.18129/B9.bioc.ropls) Bioconductor package [[@thevenot_analysis_2015]](https://doi.org/10.1021/acs.jproteome.5b00354):

```{r plsda}
sacPlsda <- ropls::opls(sacurine.eset, "gender")
```

```{r plsda_getEset}
sacurine.eset <- ropls::getEset(sacPlsda)
```

### Feature selection

With the [**biosigner**](https://doi.org/10.18129/B9.bioc.biosigner) Bioconductor package [[@rinaudo_biosigner:_2016]](https://doi.org/10.3389/fmolb.2016.00026):

```{r biosigner}
sacSign <- biosigner::biosign(sacurine.eset, "gender", seedI = 123)
sacurine.eset <- biosigner::getEset(sacSign)
```

## **annotating**: MS annotation

This method is based on the [**biodb**](https://doi.org/10.18129/B9.bioc.biosigner) R package available on github:
```{r installing_biodb, eval = FALSE}
devtools::install_github("pkrog/biodb")
```

Viewing the parameters from the **annotating** method and their default values:
```{r annotating_parameters}
phenomis::annotating_parameters()
```


### Chemical annotation with ChEBI

#### by mz values

```{r chebi_annotation, message=FALSE, warning=FALSE}
sacurine.eset <- phenomis::annotating(sacurine.eset,
                                      database.c = "chebi",
                                      param.ls = list(query.type = "mz",
                                                     query.col = "mass_to_charge",
                                                     ms.mode = "neg",
                                                     mz.tol = 10,
                                                     mz.tol.unit = "ppm",
                                                     max.results = 3,
                                                     prefix = "chebiMZ."),
                                      report.c = "none")
knitr::kable(head(Biobase::fData(sacurine.eset)[, grep("chebiMZ", Biobase::fvarLabels(sacurine.eset))]))
```

#### by ChEBI identifiers

```{r chebi_identifier, message=FALSE, warning=FALSE}
sacurine.eset <- phenomis::annotating(sacurine.eset, database.c = "chebi",
                                      param.ls = list(query.type = "chebi.id",
                                                      query.col = "database_identifier",
                                                      prefix = "chebiID."))
knitr::kable(head(Biobase::fData(sacurine.eset)[, grep("chebiID", Biobase::fvarLabels(sacurine.eset))]))

```


### Chemical annotation with a local database

Loading a local (example) MS database:
```{r localdbDF}
msdbDF <- read.table(system.file("extdata/local_ms_db.tsv", package = "phenomis"),
                     header = TRUE,
                     sep = "\t",
                     stringsAsFactors = FALSE)
```

Querying the local database
```{r localms_annotation, message=FALSE, warning=FALSE}
sacurine.eset <- phenomis::annotating(sacurine.eset,
                                      database.c = "local.ms",
                                      param.ls = list(query.type = "mz",
                                                      query.col = "mass_to_charge",
                                                      ms.mode = "neg",
                                                      mz.tol = 5,
                                                      mz.tol.unit = "ppm",
                                                      local.ms.db = msdbDF,
                                                      prefix = "localMS."),
                                      report.c = "none")
knitr::kable(Biobase::fData(sacurine.eset)[!is.na(Biobase::fData(sacurine.eset)[, "localMS.accession"]), grep("localMS.", Biobase::fvarLabels(sacurine.eset), fixed = TRUE)])
```

## **writing**: Exporting the results

```{r writing, eval = FALSE}
phenomis::writing(sacurine.eset, dir.c = getwd())
```

```{r writing_in_figures_repos, include=FALSE}
phenomis::writing(sacurine.eset, dir.c = 'figures/temp/', overwrite.l = TRUE)
```

# References
